clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2077,6 +2077,21 @@ void Sema::finalizeSYCLDelayedAnalysis(const FunctionDecl *Caller,
   }
 }
 
+bool Sema::isArrayZeroInitialized(VarDecl *VD, bool CheckValueDependent) {
+  assert(getLangOpts().SYCLIsDevice &&
+         "Should only be called during device compilation");
+  if (VD->isInvalidDecl() || !VD->hasInit() || !VD->hasGlobalStorage())
+    return false;
+  APValue *Value = nullptr;
+  if (VD->getInit() && !VD->getInit()->isValueDependent())
+    Value = VD->evaluateValue();
+  // zero-initialized arrays do not have any initialized elements.
+  if (Value && Value->isArray() && Value->hasArrayFiller() &&
```

Comment:
Are these additional checks required to handle the case `int array2[66000] = {};` @erichkeane would you know if there is a better way to check this?

Reply:
I think this is overly constrained, yes.  If all you care about is whether the array is default initialized (as this seems to be doing?), I believe you can do this without the evaluation.  You should be able to just do VD->getAnyInitializer(), which should give you an InitListExpr since this is an array type.

Then, you should be able to loop through its 'inits' to validate whether they evaluate to zero.  This would also be able to catch the cases where you get int array[6] = {0,0};, which is ALSO zero initialized, but wouldn't be caught here as far as I can tell.

Reply:
All this starts to become quite complex...
And what if the the array is not used at all and eliminated by some aggressive LLVM passes or if the code is optimized differently or...?
Here we make a lot of guesses ahead on what might happen quite later downstream in code-gen for one specific back-end...

Reply:
Typically we don't suppress warnings based on potential optimizations.  If something is illegal with NO optimizations, it also stays illegal with O3.  That is, the validity of a program should never depend on the optimization settings.

Reply:
I think we agree. But here we try to guess in Sema something that is unsupported by some LLVM back-ends... If I understand the title of th PR, it is not about warning but error.

Reply:
> I think this is overly constrained, yes. If all you care about is whether the array is default initialized (as this seems to be doing?), I believe you can do this without the evaluation. You should be able to just do VD->getAnyInitializer(), which should give you an InitListExpr since this is an array type.
> 
> Then, you should be able to loop through its 'inits' to validate whether they evaluate to zero. This would also be able to catch the cases where you get int array[6] = {0,0};, which is ALSO zero initialized, but wouldn't be caught here as far as I can tell.

Are you referring to the if checks as overly constrained here? If yes, then `Value->isArray()` had to be added because of assert checks in [getArrayInitializedElts()](https://clang.llvm.org/doxygen/APValue_8h_source.html#l00486)
 
For the case `int array[6] = {0,0};` , the first 2 elements will be set to 0 ( and the rest also set to 0) but, `Value->getArrayInitializedElts() will be equal to 2 `. This will not be a zero-initialized array and error will not be emitted for this case.
I think both the approaches  would work, so how is one better than the other?

Reply:
> 
> 
> > I think this is overly constrained, yes. If all you care about is whether the array is default initialized (as this seems to be doing?), I believe you can do this without the evaluation. You should be able to just do VD->getAnyInitializer(), which should give you an InitListExpr since this is an array type.
> > Then, you should be able to loop through its 'inits' to validate whether they evaluate to zero. This would also be able to catch the cases where you get int array[6] = {0,0};, which is ALSO zero initialized, but wouldn't be caught here as far as I can tell.
> 
> Are you referring to the if checks as overly constrained here? If yes, then `Value->isArray()` had to be added because of assert checks in [getArrayInitializedElts()](https://clang.llvm.org/doxygen/APValue_8h_source.html#l00486)

Except you already know that the type is an array, since it is a ConstantArrayType, right?

> For the case `int array[6] = {0,0};` , the first 2 elements will be set to 0 ( and the rest also set to 0) but, `Value->getArrayInitializedElts() will be equal to 2 `. This will not be a zero-initialized array and error will not be emitted for this case.

From my reading of the code, that WILL diagnose (int array[67000] = {0,0}), since it won't be zero initialized, right?  Since you'd return false in that case here.  That is distinctly the incorrect behavior.

> I think both the approaches would work, so how is one better than the other?
Because looping through the inits allows us to better early-escape, evaluating the entire initializer is a REALLY expensive operation, so we can do 'at most' the same amount of work by testing each initializer individually.  It also gives us the opportunity to deal with defaulted constructors on record types, which have some otherwise interesting behavior.

Reply:
> that WILL diagnose (int array[67000] = {0,0}), since it won't be zero initialized, right
Right, but since the size(6) is < 65532, the error will not be emitted.

Reply:
Note the example you quoted.  I did a larger array to show that your exception here based on zero-initialized is broken.  I was using the '6' simply to be different from the element count (and not 0).

If :
int array[67000] = {}; 
Shouldn't diagnose, neither should:
int array[67000] = {0,0,{},0};

Nor should:

DefaultConstructible DC[67000] = {};
or
DefaultConstructible DC[67000] = {{}, DefaultConstructible{}};


There is question whether
EmptyDefaultConstructible DC[67000] = {};
or
EmptyDefaultConstructible DC[67000] = {{}, EmptyDefaultConstructible{}};

Should be diagnosed.